text/0000-futures.md

```diff
@@ -154,31 +164,316 @@ can be dropped.
+executor-specific `RawWaker`.
+
+This mechanism is chosen in favor of trait objects since it allows for more
+flexible memory management schemes. `RawWaker` can be implemented purely in
+terms of global functions and state, on top of reference counted objects, or
+in other ways.
+
+The relation between those `Waker` types is outlined in the following definitions:
+
+```rust
+/// A `RawWaker` allows the implementor of a task executor to customize the
+/// behavior of `LocalWaker`s and `Waker`s.
+/// It consists of a data pointer a virtual function pointer table (vtable) that
+/// customizes the behavior of the `RawWaker`.
+#[derive(PartialEq)]
+pub struct RawWaker {
+    /// A data pointer, which can be used to store arbitrary data as required
+    /// by the executor. This could be e.g. a type-erased pointer to an `Arc`
+    /// that is associated with the task.
+    /// The value of this field gets passed to all functions that are part of
+    /// the vtable as first parameter.
+    pub data: *const (),
+    /// Virtual function pointer table that customizes the behavior if this waker.
+    pub vtable: &'static RawWakerVTable,
+}
+
+/// A virtual function pointer table (vtable) that allows to customize the
+/// behavior of a `RawWaker`.
+/// The pointers which get passed to all functions inside the vtable are the
+/// values of the `data` field.
+#[derive(PartialEq, Copy, Clone)]
+pub struct RawWakerVTable {
+    /// This function will be called when the `RawWaker` gets cloned, e.g. when
+    /// the `Waker` or `LocalWaker` in which the `RawWaker` is stored gets cloned.
+    /// The implementation of this function must retain all resources that are
+    /// required for this additional instance of a `RawWaker` and associated
+    /// task.
+    pub clone: unsafe fn(*const ()) -> RawWaker,
+    /// This function will be called when a `LocalWaker` should be converted into
+    /// a thread-safe `Waker`. The implementation of this function must return
+    /// a new `RawWaker` which can fulfill the requirements of `Waker`. It can
+    /// exchange the vtable for that purpose. E.g. it might replace the `wake`
+    /// function with a varient that supports cross-thread wakeups.
+    ///
+    /// The old `LocalWaker` which contained the data pointer should be consumed
+    /// while performing the operation. After the operation the `LocalWaker`
+    /// won't exist anymore, only the new `Waker`.
+    /// This means that if both instances would utilize the same reference-counted
+    /// data, changing the reference count would not be necessary.
+    ///
+    /// If a conversion is not supported, the implementation should return
+    /// `None`. In this case the implementation must still make
+    /// sure that the data pointer which is passed to the function is correctly
+    /// released, e.g. by calling the associated `drop_fn` on it.
+    pub into_waker: unsafe fn(*const ()) -> Option<RawWaker>,
```

Comment:
I don't think we should return an `Option` here-- the only reasonable behavior on `None` when you need a `Send`-able waker is to panic, so I think the panic-ing should be left up to the implementation of `into_waker`.

Reply:
The publicly visible API (`LocalWaker`) has 2 methods:
- `into_waker()` which panics when this delegate returns `None`.
- `try_into_waker()` which forwards the result.

I'm not sure about too many use cases where `try_into_waker()` would be useful apart from maybing printing a compatibility error. But it doesn't really hurt to expose it. And I felt that having the `panic` at a central place instead of in the actual waker implementations seems cleaner. Raises e.g. less questions if the panic happened before or after the old `RawWaker` was consumed.

Reply:
Makes sense! thanks for the explanation